docs/developer_guide/add_a_decoder.md

```diff
@@ -1,46 +1,89 @@
 # 1. Add a new decoder class
 
 Source code for decoders lives under `ludwig/decoders/`.
-New decoder objects should be defined in the corresponding files, for example all new sequence decoders should be added to `ludwig/decoders/sequence_decoders.py`.
+Decoders are grouped into modules by their output feature type. For instance, all new sequence decoders should be added
+to `ludwig/decoders/sequence_decoders.py`.
 
-All the decoder parameters should be provided as arguments in the constructor with their default values set.
-For example the `SequenceGeneratorDecoder` decoder takes the following list of arguments in its constructor:
+!!! note
+
+    A decoder may support multiple output types, if so it should be defined in the module corresponding to its most
+    generic supported type. If a decoder is generic with respect to output type, add it to
+    `ludwig/decoders/generic_decoders.py`.
+
+To create a new decoder:
+
+1. Define a new decoder class. Inherit from `ludwig.decoders.base.Decoder` or one of its subclasses.
+2. Create all layers and state in the `__init__` method, after calling `super().__init__()`.
+3. Implement your decoder's forward pass in `def forward(self, inputs, **kwargs):`.
+4. Define `@property input_shape`.
```

Comment:
This is actually not trivial to define for decoders since the input to the decoder is a dictionary of tensors. Sequence decoders, for now, return a dummy shape.